File: automata/automata.js
// record itteration
let iteration = 0;

// Returns a 2 dimensional, 0 filled array, based on a specified width and height.
function makeWorld(w, h) {
  const world = new Array(h).fill(null);
  world.forEach((item, i) => {
    world[i] = new Array(w).fill(0);
  });
  return world;
}

// Gets the current world, runs a generation, and returns the next world.
function evolve(world) {
  // init res
  const res = makeWorld(world[0].length, world.length);

  // set res to world without refrencing problems.
  world.forEach((row, y) => {
    world[y].forEach((point, x) => {
      res[y][x] = world[y][x];
    });
  });

  // record which collum has growed here
  const collumGrowed = new Array(world[0].length).fill(false);

  // run on every pixel
  world.forEach((row, y) => {
    world[y].forEach((point, x) => {
      // if not on edge
      if (y > 0 && y < world.length && x > 0 && x < world[y].length) {
        // if cell is living, and collum has not grown yet.
        if (world[y][x] === 1 && collumGrowed[x] === false) {
          // This collum has grown
          collumGrowed[x] = true;

          // calculate chance for horizontal movement
          const horzChance = (world.length - 1 - y) / 30;

          // witch direction to grow
          // there is a [horzchance]% chance to grow right or left
          // there is a [(1 - horzChance) * 0.125]% chance to grow up
          if (Math.random() < horzChance) {
            res[y][x + 1] = 1;
          }
          if (Math.random() < horzChance) {
            res[y][x - 1] = 1;
          }

          if (Math.random() < (1 - horzChance) * 0.125) {
            res[y - 1][x] = 1;
          }
        }
      }
    });
  });

  return res;
}

// render world
function render(world) {
  world.forEach((row, y) => {
    const str = row.join('').split('1').join('█').split('0')
      .join(' ');
    const colors = [];
    if (y > world.length - 5) {
      colors.push('\x1b[37m');
      colors.push('\x1b[0m');
    } else {
      colors.push('\x1b[32m');
      colors.push('\x1b[0m');
    }
    console.log(colors[0], str, colors[1]); // eslint-disable-line no-console
  });
}

// animate world
function animate(interval, input, times) {
  iteration += 1;
  let world = input;
  world = evolve(world);
  render(world);

  if (iteration < times || times === undefined) {
    setTimeout(() => {
      animate(interval, world, times);
    }, interval);
  }
}

// make a world
const world = makeWorld(process.stdout.columns - 2, process.stdout.rows - 1);

for (let i = 1; i < (parseInt(process.argv[2], 10)) + 1; i += 1) {
  const y = world.length - 1;
  const x = Math.floor(world[0].length / (parseInt(process.argv[2], 10) + 1)) * i;
  world[y][x] = 1;
}

if (process.argv[2] !== undefined) {
  animate(100, world, 90);
} else {
  console.log('Error: please give a number'); // eslint-disable-line no-console
}

File: automata/automata.py
import random
import time
import os
import sys

arguments = sys.argv

# record itteration
iteration = 0

def make_world(w, h):
  return [[0 for x in range(w)] for y in range(h)]

def evolve(world):

  # init res
  res = make_world(len(world[0]), len(world))

  # set res to world without refrencing problems.
  for y in range(len(world)):
    for x in range(len(world[y])):
      res[y][x] = world[y][x]

  # record which collum has growed here
  collumGrowed = [0] * len(world[0])

  for y in range(len(world)):
    for x in range(len(world[y])):
      # if cell is living, the collum has not grown yet, and not on edge.
      if y > 0 and y < len(world) and x > 0 and x < len(world[y]) - 1 and world[y][x] == 1 and collumGrowed[x] == 0:
        # this collum has grown
        collumGrowed[x] = 1

        # calculate chance for horizontal movement
        horzChance = float(len(world) - 1 - y) / 30
        if random.random() < horzChance:
          res[y][x + 1] = 1;

        if random.random() < horzChance:
          res[y][x - 1] = 1;

        if random.random() < ((1 - horzChance) * 0.125):
          res[y - 1][x] = 1;

  return res


def render(world):
  for y in range(len(world)):
    string = ''
    for x in range(len(world[y])):
      string = string + str(world[y][x]).replace('0', ' ').replace('1', 'x')
    if(y > len(world) - 5):
      print('\x1b[37m' + string)
    else:
      print('\x1b[32m' + string)

  print(iteration)


height, width = os.popen('stty size', 'r').read().split()
w = make_world(int(width), int(height) - 1)


if not len(sys.argv) < 2:
  n = int(sys.argv[1]) + 1
  for x in xrange(1,n):
    w[len(w) - 1][int(len(w[0]) / n) * x] = 1


  for x in xrange(0,90):
    iteration += 1
    time.sleep(0.1)
    w = evolve(w)
    render(w)
else:
  print('Error: please give a number')
